onnx/shape_inference/implementation.cc

```diff
@@ -166,23 +166,32 @@ static void InferShapesImpl(
     ) {
   std::unordered_map<std::string, TypeProto*> valueTypesByName{
       outer_scope_value_types_by_name};
+  std::unordered_map<std::string, TypeProto*> undefinedValueTypesByName{
+      outer_scope_value_types_by_name};
 
   GraphInferenceContext graphInferenceContext{
       valueTypesByName, opset_imports, schema_registry};
 
   for (auto& vi : *g->mutable_value_info()) {
-    if (vi.has_type())
+    if (vi.has_type()) {
       valueTypesByName[vi.name()] = vi.mutable_type();
+    }
   }
   for (auto& vi : *g->mutable_input()) {
-    if (vi.has_type())
+    if (vi.has_type()) {
       valueTypesByName[vi.name()] = vi.mutable_type();
+    }
   }
   for (auto& vi : *g->mutable_output()) {
-    // Some output type might be undefined
-    // To assgin inferred type to them,
-    // Also save names of output with undefined types
-    valueTypesByName[vi.name()] = vi.mutable_type();
+    // TODO: Remove this condition and undefinedValueTypesByName in the future
```

Comment:
either remove this comment or explain why this condition and undefinedValueTypesByName  needs to be removed. My suggestion remove this comment from here and create an issue to track this item. The first step is to add the clarification in the spec and then update the code here.

Reply:
Sounds good. I have removed it and tracked here: https://github.com/onnx/onnx/issues/3069